chia/data_layer/download_data.py

```diff
@@ -1,166 +1,159 @@
-            for row in answer:
-                if row["is_terminal"]:
-                    key = row["key"]
-                    value = row["value"]
-                    hash = Program.to((hexstr_to_bytes(key), hexstr_to_bytes(value))).get_tree_hash()
-                    if hash.hex() == node:
-                        insert_batch.append((NodeType.TERMINAL, key, value))
-                        right_hash = hash.hex()
-                        while right_hash in add_to_db_cache:
-                            node, left_hash = add_to_db_cache[right_hash]
-                            del add_to_db_cache[right_hash]
-                            insert_batch.append((NodeType.INTERNAL, left_hash, right_hash))
-                            right_hash = node
-                    else:
-                        raise RuntimeError(f"Did not received expected node. Expected: {node} Received: {hash.hex()}")
-                    if len(stack) > 0:
-                        node = stack.pop()
-                    else:
-                        node = None
-                else:
-                    left_hash = row["left"]
-                    right_hash = row["right"]
-                    left_hash_bytes = bytes32.from_hexstr(left_hash)
-                    right_hash_bytes = bytes32.from_hexstr(right_hash)
-                    hash = Program.to((left_hash_bytes, right_hash_bytes)).get_tree_hash(
-                        left_hash_bytes, right_hash_bytes
-                    )
-                    if hash.hex() == node:
-                        add_to_db_cache[right_hash] = (node, left_hash)
-                        # At most max_height nodes will be pending to be added to DB.
-                        assert len(add_to_db_cache) <= 100
-                    else:
-                        raise RuntimeError(f"Did not received expected node. Expected: {node} Received: {hash.hex()}")
-                    stack.append(right_hash)
-                    node = left_hash
-
-            if node is not None:
-                raise RuntimeError("Did not download full data.")
-            await ws.send_str(json.dumps({"type": "close"}))
-
-    await data_store.insert_batch_for_generation(
-        insert_batch, tree_id, bytes32.from_hexstr(root_json["node_hash"]), int(root_json["generation"]), lock=lock
-    )
-    return True
+    if root.node_hash is not None:
+        node_hash = root.node_hash
+    else:
+        node_hash = bytes32([0] * 32)  # todo change
+    filename_full_tree = foldername.joinpath(get_full_tree_filename(tree_id, node_hash, root.generation))
+    filename_diff_tree = foldername.joinpath(get_delta_filename(tree_id, node_hash, root.generation))
+    written = False
+    if override and filename_full_tree.exists():
+        os.remove(filename_full_tree)
+    if not filename_full_tree.exists():
+        await data_store.write_tree_to_file(root, node_hash, tree_id, False, filename_full_tree)
+        written = True
```

Comment:
This still has a race condition where even with an override it can still just not write anything if the file is recreated after we delete it.  I think we can open with `w` if overriding and `x` and handle the _specific_ exception and number.  Maybe double check the truncation behavior with `x`.

Same below.

```pycon
>>> f = open("abcd", "x")
>>> f = open("abcd", "x")
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
FileExistsError: [Errno 17] File exists: 'abcd'
```